pkg/ipam/poolallocator/allocator.go

```diff
@@ -272,6 +309,63 @@ func (a *IPPoolAllocator) AllocateNext(state v1alpha2.IPAddressPhase, owner v1al
 	return ip, subnetSpec, err
 }
 
+// AllocateReserved allocates the reserved IP if exists, else allocates next
+// available IP. It returns error if pool is exausted, or in case CRD failed to update its state.
+// In case of success, IP pool CRD status is updated with allocated IP/state/resource/container.
+// AllocateIP returns subnet details for the requested IP, as defined in IP pool spec.
+func (a *IPPoolAllocator) AllocateReserved(state v1alpha2.IPAddressPhase, owner v1alpha2.IPAddressOwner) (net.IP, v1alpha2.SubnetInfo, error) {
+	var subnetSpec v1alpha2.SubnetInfo
+	var ip net.IP
+
+	ip, err := a.CheckReserved(owner)
+	if err != nil {
+		return ip, subnetSpec, err
+	}
+	if ip == nil {
+		// ip is not reserved, allocate next available ip
+		return a.AllocateNext(state, owner)
+	}
+
+	// Same resource can not ask for allocation twice without release
+	// This needs to be verified even at the expence of another API call
+	exists, err := a.HasContainer(owner.Pod.ContainerID)
+	if err != nil {
+		return ip, subnetSpec, err
+	}
+	if exists {
+		return ip, subnetSpec, fmt.Errorf("container %s was already allocated an address from IP Pool %s", owner.Pod.ContainerID, a.ipPoolName)
+	}
+
+	// Retry on CRD update conflict which is caused by multiple agents updating a pool at same time.
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ipPool, allocators, err := a.getPoolAndInitIPAllocators()
+		if err != nil {
+			return err
+		}
+
+		index := len(allocators)
```

Comment:
I think normally -1 is used to indicate invalid index

Reply:
Changed to `-1` as the initial/invalid value.